### PR TITLE
Implement workspace and document cache

### PR DESCRIPTION
**Pull request by msujew**
Implements a feature similar to the Xtext `ResourceScopeCache`.

### CODE REVIEW

File: packages/langium/src/utils/caching.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    get(documentUri: URI, key: K): V | undefined;
+    delete(documentUri: URI, key: K): boolean;
+    clear(): void;
+    clear(documentUri: URI): void;
+}
+
+export interface WorkspaceCache<K, V> extends Disposable {
+    has(key: K): boolean;
+    set(key: K, value: V): void;
+    get(key: K): V | undefined;
```
Comment (spoenemann) [inline packages/langium/src/utils/caching.ts:29]: The most obvious pattern of a cache is
```typescript
if (cache.has(key)) {
    return cache.get(key);
} else {
    const value = computeValue();
    cache.set(key, value);
    return value;
}
```
IMO every cache implementation should support this pattern, e.g. like this:
```suggestion
    get(key: K, valueProvider: () => V): V;
    get(key: K): V | undefined;
```
or move this to another method name like `access`.

File: packages/langium/src/services.ts [hunk 2]
```diff
@@ -158,6 +159,7 @@ export type LangiumGeneratedSharedServices = {
  */
 export type LangiumDefaultSharedServices = {
     ServiceRegistry: ServiceRegistry
+    Caching: CachingService
```
Comment (spoenemann) [inline packages/langium/src/services.ts:162]:
```suggestion
    CachingService: CachingService
```

**Review by spoenemann — COMMENTED** [review]
Thanks @msujew! Could you add a proposal about how to use the new service to the PR? We should have a clear picture of where such Map-based caches could help to improve performance.

AFAICT #1121 could make use of a WorkspaceCache in the scope provider, right?

**Comment by msujew** [comment]
> AFAICT https://github.com/langium/langium/pull/1121 could make use of a WorkspaceCache in the scope provider, right?

Yes, exactly :)

**Comment by spoenemann** [comment]
I think it would make sense to integrate these changes into #1121 then, to see the new utility in action and to clean up the changes made there.

**Review by spoenemann — COMMENTED** [review]
I propose a different class structure to make the underlying cache implementation more reusable and reduce the necessary API surface:
 - `SimpleCache<K, V>` as a simple wrapper around `Map`
 - `ContextCache<C, K, V>` with a two-stage map
 - `WorkspaceCache<K, V>` (without interface) extends `SimpleCache<K, V>`, creates and disposes the DocumentBuilder listener internally (receives LangiumSharedServices in constructor)
 - `DocumentCache<K, V>` (without interface) extends `ContextCache<string, K, V>`, creates and disposes the DocumentBuilder listener internally (receives LangiumSharedServices in constructor)
 - Dispose the `DisposableCache` superclass
 - `CachingService` is no longer necessary because you can easily create the caches in your service constuctor:
```typescript
class MyService {
    private myCache: DocumentCache<A, B>;
    constructor(services: MyLanguageServices) {
        this.myCache = new DocumentCache(services.shared);
    }
}
```
- If you need a short-living cache, create a factory in your constructor:
```typescript
class MyService {
    private myCacheFactory: () => DocumentCache<A, B>;
    constructor(services: MyLanguageServices) {
        this.myCacheFactory = () => new DocumentCache(services.shared);
    }
}
```

File: packages/langium/src/utils/caching.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    dispose(): void {
+        this.throwIfDisposed();
+        this.clear();
+        this.isDisposed = true;
+        this.toDispose.dispose();
+    }
+
+    throwIfDisposed(): void {
```
Comment (spoenemann) [inline packages/langium/src/utils/caching.ts:27]: This should be protected.

File: packages/langium/src/utils/caching.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+import type { Disposable } from 'vscode-languageserver';
+import type { URI } from 'vscode-uri';
+import type { LangiumSharedServices } from '../services';
+
+export abstract class DisposableCache implements Disposable {
+
+    protected toDispose: Disposable;
+    protected isDisposed = false;
+
+    constructor(disposable: Disposable) {
```
Comment (spoenemann) [inline packages/langium/src/utils/caching.ts:16]: Proposal: add a method `onDispose` that adds a disposable to an array to separate this from construction.

File: packages/langium/src/utils/caching.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        this.cache.clear();
+    }
+}
+
+export class ContextCache<I, C, K, V> extends DisposableCache {
+
+    private cache = new Map<C, Map<K, V>>();
+    private converter: (input: I) => C;
+
+    constructor(disposable: Disposable, converter: (input: I) => C) {
```
Comment (spoenemann) [inline packages/langium/src/utils/caching.ts:80]: `converter` could be optional for `I` = `C`?

File: packages/langium/src/utils/caching.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    clear(contextKey?: I): void {
+        this.throwIfDisposed();
+        if (contextKey) {
+            this.getDocumentCache(contextKey).clear();
+        } else {
+            this.cache.clear();
+        }
+    }
+
+    protected getDocumentCache(contextKey: I): Map<K, V> {
```
Comment (spoenemann) [inline packages/langium/src/utils/caching.ts:127]:
```suggestion
    protected getCacheForContext(contextKey: I): Map<K, V> {
```

**Review by spoenemann — COMMENTED** [review]
This looks really good! More detailed comments:

**Pull request merged**